src/Compilers/Core/Portable/SourceGeneration/Nodes/InputNode.cs

```diff
@@ -56,10 +57,21 @@ public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, N
                     var usedCache = builder.TryUseCachedEntries();
                     Debug.Assert(usedCache);
                 }
+                else if (inputItems.Length == previousTable.Count)
+                {
+                    // When the number of items matches the previous iteration, we use a heuristic to mark the input as modified
+                    // This allows us to correctly 'replace' items even that aren't actually the same. In the case that the item 
+                    // really isn't modified, but a new item, we still function correctly as we mostly treat them the same, but
+                    // will perform an extra comparison that is ommitted in the pure 'added' case.
+                    var modified = builder.TryModifyEntry(inputItems[itemIndex], _comparer);
```

Comment:
>TryModifyEntry(inputItems[itemIndex], _comparer) [](http://example.com/codeflow?start=43&length=48)

What if this item from `inputItems` appears later in `previousTable`? Won't we add it twice to the new state?

Reply:
Then we'll lookup that item in the `itemsSet` and find that its no longer there. We'll go into the heuristic block and end up marking the item as having been modified but replaced with the equivalent `itemIndex` from the inputs. In those cases we might accidentally say some inputs where modified, when they were really added/removed, but that just means we do a bit more work later on (by comparing the values).

Reply:
If two input items are swapped but otherwise identical to the original two, won't we mark one of the two as new rather than marking both as cached?